cli/bench/lsp.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::error::generic_error;
+use deno_core::error::AnyError;
+use deno_core::serde::de;
+use deno_core::serde::Deserialize;
+use deno_core::serde_json;
+use deno_core::serde_json::json;
+use deno_core::serde_json::Value;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::process::Stdio;
+use tokio::io::AsyncBufReadExt;
+use tokio::io::AsyncReadExt;
+use tokio::io::AsyncWriteExt;
+use tokio::io::BufReader;
+use tokio::io::BufWriter;
+use tokio::io::Lines;
+use tokio::process::ChildStderr;
+use tokio::process::ChildStdin;
+use tokio::process::ChildStdout;
+use tokio::process::Command;
+use tokio::runtime;
+use tokio::time::Instant;
+
+lazy_static! {
+  static ref CONTENT_TYPE_REG: Regex =
+    Regex::new(r"(?i)^content-length:\s+(\d+)").unwrap();
+}
+
+#[derive(Debug, Deserialize)]
+struct LspResponse {
+  pub result: Option<Value>,
+  pub id: u32,
+  pub error: Option<LspResponseError>,
+}
+
+#[derive(Debug, Deserialize)]
+struct LspResponseError {
+  pub code: i32,
+  pub message: String,
+  pub data: Option<Value>,
+}
+
+#[derive(Debug, Deserialize)]
+struct LspNotification {
+  pub method: String,
+  pub params: Option<Value>,
+}
+
+fn create_basic_runtime() -> runtime::Runtime {
+  runtime::Builder::new_current_thread()
+    .enable_io()
+    .enable_time()
+    // for performance purposes, we might as well allow this to go to as
+    // many threads as it can, so that the harness doesn't become the bottle
+    // neck
+    // .max_blocking_threads(32)
+    .build()
+    .unwrap()
+}
+
+pub(crate) fn run_local<F, R>(future: F) -> R
+where
+  F: std::future::Future<Output = R>,
+{
+  let rt = create_basic_runtime();
+  rt.block_on(future)
+}
+
+struct LspClient {
+  pub output: Lines<BufReader<ChildStderr>>,
+  reader: BufReader<ChildStdout>,
+  start: Instant,
+  writer: BufWriter<ChildStdin>,
+}
+
+impl LspClient {
+  pub fn new(deno_exe: &PathBuf) -> Result<Self, AnyError> {
+    let mut child = Command::new(deno_exe)
+      .kill_on_drop(true)
```

Comment:
Aside: `kill_on_drop` - cool - I didn't know about this method

Reply:
Only on `tokio::process::Command`...

Reply:
Ah I didn't realize this was tokio::process::Command... Do we really need to interact with the "deno lsp" child process in an asynchronous way? I think std::process::Command might be sufficient.  I'm slightly worried we might end up running multiple benchmarks concurrently or something.

Reply:
I have found it really hard to find a way to not block the stdout/stdin, and if they get blocked, the lsp can stop responding, it was the problem I had when I first tried the integration harness, which is partly why I was avoiding revisiting it, but the `tokio::process::Command` makes it so much easier to not block, and we really need to handle the requests and responses.  I actually found out that if I fired in all the commands and then read the results, the lsp actually complains that the "shutdown" command is invalid, because it knows that other messages haven't been sent.

Not going to run multiple benchmarks at the same time.  That would only happen with multiple instances of the `LspClient` which simply won't occur and the collection of the benchmarks for the lsp is blocked in the main.rs.

Reply:
As far as I can see below, you await each read/write serially. AFAICT there would be no effective difference if this was done with blocking syscalls.